OvmfPkg/QemuKernelLoaderFsDxe/QemuKernelLoaderFsDxe.c

```diff
@@ -291,6 +291,7 @@ QemuKernelBlobTypeToFileInfo (
 
   NameSize     = (StrLen (Name) + 1) * 2;
   FileInfoSize = OFFSET_OF (EFI_FILE_INFO, FileName) + NameSize;
+  ASSERT (FileInfoSize >= sizeof *FileInfo);

```

Comment:
use of sizeof() on a structure that uses an array of [1] at the end to indicate a flexible array member is not correct.  Instead, the OFFSET_OF() macro should be used to determine the size of a structure declared in this way.

Of course proper use of flexible array member would be much cleaner.  However, some of the UEFI Spec structures were defined before flexible array members would available in all compilers and C standards.

Reply:
@mdkinney Hi Mike, `sizeof *FileInfo` is intentional here. The ASSERT() expresses that the invariable parts of `EFI_FILE_INFO`, plus the variable/flexible ("struct hack") `FileName` mmber, are *not shorter* than the size of `EFI_FILE_INFO`. In other words, that `EFI_FILE_INFO` does not get truncated due to a short `FileName`.

> Of course proper use of flexible array member would be much cleaner. However, some of the UEFI Spec structures were defined before flexible array members would available in all compilers and C standards.

This is (independently) highly relevant, as the UEFI spec actually declares `EFI_FILE_INFO.FileName` as a (C99) flexible array member, but edk2 declares the same member as a 1-element array (C89 struct hack).